issue_6/contracts/GreenAMM.sol

```diff
@@ -0,0 +1,360 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.8.4;
+
+import "@yield-protocol/utils-v2/contracts/access/Ownable.sol";
+import "@yield-protocol/utils-v2/contracts/token/ERC20.sol";
+
+/**
+@dev Internal library that implements Fixed Point Math
+WADs are fixed point numbers with decimals=18
+ */
+library FPMath {
+    // /**
+    // @notice Multiply 2 WADs
+    // */
+    uint256 internal constant WAD = 1e18;
+
+    function wadMul(uint256 a, uint256 b) internal pure returns (uint256) {
+        return (a * b) / WAD;
+    }
+
+    /// @dev div 2 WADs
+    function wadDiv(uint256 a, uint256 b) internal pure returns (uint256) {
+        return (a * WAD) / b;
+    }
+}
+
+/// @dev placeholder to keep all token-related data
+struct TokenData {
+    uint256 reserves;
+    // ended up not needing anything else
+}
+
+/**
+@notice 'Engine' of the standard xy=k AMM contract
+ */
+abstract contract GreenAMMEngine is Ownable {
+    bool internal isInitialized;
+
+    /// @dev info about X token
+    TokenData public x;
+    /// @dev info about Y token
+    TokenData public y;
+
+    /// @dev K from X*Y=K
+    /// IMIPORTANT: K has 36 decimals
+    uint256 public k;
+
+    event KChanged(uint256 newK);
+    event Inited(uint256 x, uint256 y);
+    event Minted(uint256 x, uint256 y);
+    event Burned(uint256 x, uint256 y);
+    event SoldX(uint256 x, uint256 y);
+    event SoldY(uint256 x, uint256 y);
+
+    modifier initialized() {
+        require(isInitialized, "not initialized");
+        _;
+    }
+
+    modifier notInitialized() {
+        require(!isInitialized, "already initialized");
+        _;
+    }
+
+    /// @dev asserts that amountX/amountY have same ratio as existing reserves
+    modifier maintainsBalance(uint256 amountX, uint256 amountY) {
+        require(x.reserves != 0 && y.reserves != 0, "Empty pool");
+
+        // since decimals for X and Y are unchanged, don't bother computing *exact* value for x/y:
+        // x1 * dec_x / y1 * dec_y == x2 * dec_x / y2 * dec_y IMPLIES x1/y1 == x2/y2
+        // x1/y1 == x2/y2 IS EQUIVALENT to x1 * y2 == x2 * y1
+        require(x.reserves * amountY == y.reserves * amountX, "Unbalancing");
+        _;
+    }
+
+    /**
+    @notice constructor
+    @param xDecimals decimals of X token
+    @param yDecimals decimals of Y token
+
+    Only 18-decimal tokens are supported
+     */
+    constructor(uint8 xDecimals, uint8 yDecimals) {
```

Comment:
It would be cleaner if you would take two IERC20Metadata parameters in the constructor.

Reply:
yes and no - it breaks the 'no external calls' rule for the Engine, and switching to addresses will blow up the search space for SMTChecker